numba/cpython/builtins.py

```diff
@@ -928,3 +929,76 @@ def impl(obj, name, default):
         return resolve_getattr(obj, name, default)
     return impl
 
+
+@intrinsic
+def resolve_hasattr(tyctx, obj, name):
+    if not isinstance(name, types.StringLiteral):
+        raise RequireLiteralValue("argument 'name' must be a literal string")
+    lname = name.literal_value
+    fn = tyctx.resolve_getattr(obj, lname)
+    # Whilst technically the return type could be a types.bool_, the literal
+    # value is resolvable at typing time. Propagating this literal information
+    # into the type system allows the compiler to prune branches based on a
+    # hasattr predicate. As a result the signature is based on literals. This is
+    # "safe" because the overload requires a literal string so each will be a
+    # different variant of (obj, literal(name)) -> literal(bool).
+    if fn is None:
+        retty = types.literal(False)
+    else:
+        retty = types.literal(True)
+    sig = retty(obj, name)
+    def impl(cgctx, builder, sig, ll_args):
+       return cgutils.false_bit if fn is None else cgutils.true_bit
+    return sig, impl
+
+# hasattr cannot be implemented as a getattr call and then catching
+# AttributeError because Numba doesn't support catching anything other than
+# "Exception", so lacks the specificity required. Instead this implementation
+# tries to resolve the attribute via typing information and returns True/False
+# based on that.
+@overload(hasattr)
+def ol_hasattr(obj, name):
+    def impl(obj, name):
+        return resolve_hasattr(obj, name)
+    return impl
+
+
+def _get_py_repr_and_warn(obj):
+    # Gets the repr of an object at runtime whilst issuing a performance
+    # warning
+    from numba import typeof
+    msg = ("There is no '__repr__' specialisation defined for the type "
+           f"'{typeof(obj)}', as a result the Python interpreter has been "
+           "called to fetch a repr(), this has a performance penalty. To avoid "
+           "this penalty either define a '__repr__' for the type or remove the "
+           "call to str()/repr()!")
+    warnings.warn(NumbaPerformanceWarning(msg))
+    return repr(obj)
+
+
+@overload(repr)
+def ol_repr_generic(obj):
+    from numba import objmode
+    def impl(obj):
+        attr = '__repr__'
+        if hasattr(obj, attr) == True:
+            return getattr(obj, attr)()
+        else:
+            # There's no __str__ or __repr__ defined for this object, go via
+            # the interpreter.
+            with objmode(pyrepr='unicode_type'):
+                pyrepr = _get_py_repr_and_warn(obj)
+                return pyrepr
```

Comment:
TODO: Need to decide about this! Another option is to raise.

Reply:
one more option is to print something like `<object XYZType>` so it doesn't need objmode

Reply:
From Numba public meeting discussion on 2023-01-10, the use of `objmode` is going to be replaced with a simple compiled in string cf. https://github.com/numba/numba/pull/8442#discussion_r1029956563. This is because users are potentially going to `str()` or `repr()` objects to essentially stringify them for use elsewhere and reacquiring the GIL in such a case would be an unnecessary performance penalty.

Reply:
This is implemented and tested in b819f01.